Optional struct reference produces a C error

### Describe the bug

This code produces a C error. `?One` and `&One` types both work fine.
```v
struct One {
    thing string = 'thing 1'
}
struct Two {
    maybe_one ?&One
}
fn main() {
    println(Two{})
}
```

### Expected Behavior

Expected to compile and print
```v
Two{
    maybe_one: Option(error: none)
}
```

### Current Behavior

Produces a C error
```
code.v:8:11: warning: reference field `Two.maybe_one` must be initialized
    6 | }
    7 | fn main() {
    8 |   println(Two{})
      |           ~~~~~
    9 | }
/tmp/v_60000/code.5494600477391241273.tmp.c:2118: error: cannot convert 'struct _option_main__One_ptr' to 'void *'
builder error: 
==================
C error. This should never happen.
```

### Reproduction Steps

Create an optional reference field. `struct Test { field ?&u8 }` this is the quickest way to reproduce

### Possible Solution

_No response_

### Additional Information/Context

_No response_

### V version

V 0.3.3 0ba0fb2.ddcc22f

### Environment details (OS name and version, etc.)

OS: windows, Microsoft Windows 11 Pro v22621 64-bit
Processor: 16 cpus, 64bit, little endian,
CC version: Error: exec failed (CreateProcess) with code 2: The system cannot find the file specified.
 cmd: cc --version

getwd: C:\Users\imado
vmodules: C:\Users\imado\.vmodules
vroot: C:\Users\imado\v
vexe: C:\Users\imado\v\v.exe
vexe mtime: 2023-03-11 19:44:33
is vroot writable: true
is vmodules writable: true
V full version: V 0.3.3 0ba0fb2.ddcc22f

Git version: git version 2.33.1.windows.1
Git vroot status: weekly.2023.09-73-gddcc22fe (11 commit(s) behind V master)
.git/config present: true
thirdparty/tcc status: thirdparty-windows-amd64 1e6e7c6f

--- a/vlib/v/gen/c/auto_str_methods.v
+++ b/vlib/v/gen/c/auto_str_methods.v
@@ -56,7 +56,11 @@ fn (mut g Gen) get_str_fn(typ ast.Type) string {
 	$if trace_autostr ? {
 		eprintln('> get_str_fn: ${typ.debug()}')
 	}
-	mut unwrapped := g.unwrap_generic(typ).set_nr_muls(0).clear_flag(.variadic)
+	mut unwrapped := if typ.has_flag(.option) {
+		g.unwrap_generic(typ).clear_flag(.variadic)
+	} else {
+		g.unwrap_generic(typ).set_nr_muls(0).clear_flag(.variadic)
+	}
 	if g.pref.nofloat {
 		if typ == ast.f32_type {
 			unwrapped = ast.u32_type
@@ -515,6 +519,9 @@ fn styp_to_str_fn_name(styp string) string {
 
 // deref_kind returns deref, deref_label
 fn deref_kind(str_method_expects_ptr bool, is_elem_ptr bool, typ ast.Type) (string, string) {
+	if typ.has_flag(.option) {
+		return '', ''
+	}
 	if str_method_expects_ptr != is_elem_ptr {
 		if is_elem_ptr {
 			return '*'.repeat(typ.nr_muls()), '&'.repeat(typ.nr_muls())
@@ -936,7 +943,11 @@ fn (mut g Gen) gen_str_for_struct(info ast.Struct, styp string, typ_str string, 
 			caller_should_free = false
 		} else if ftyp_noshared.is_ptr() {
 			// reference types can be "nil"
-			funcprefix += 'isnil(it.${c_name(field.name)})'
+			if ftyp_noshared.has_flag(.option) {
+				funcprefix += 'isnil(&it.${c_name(field.name)})'
+			} else {
+				funcprefix += 'isnil(it.${c_name(field.name)})'
+			}
 			funcprefix += ' ? _SLIT("nil") : '
 			// struct, floats and ints have a special case through the _str function
 			if sym.kind !in [.struct_, .alias, .enum_, .sum_type, .map, .interface_]

--- a/vlib/v/gen/c/str.v
+++ b/vlib/v/gen/c/str.v
@@ -125,7 +125,7 @@ fn (mut g Gen) gen_expr_to_string(expr ast.Expr, etype ast.Type) {
 		if str_method_expects_ptr && !is_ptr {
 			g.write('&')
 		} else if is_ptr && typ.has_flag(.option) {
-			g.write('*(${g.typ(typ.set_nr_muls(0))}*)&')
+			g.write('*(${g.typ(typ)}*)&')
 		} else if !str_method_expects_ptr && !is_shared && (is_ptr || is_var_mut) {
 			g.write('*'.repeat(typ.nr_muls()))
 		}

--- a/vlib/v/tests/option_print_ptr_test.v
+++ b/vlib/v/tests/option_print_ptr_test.v
@@ -1,0 +1,14 @@
+struct One {
+	thing string = 'thing 1'
+}
+
+struct Two {
+	maybe_one ?&One
+}
+
+fn test_dump_option_ptr() {
+	a := Two{}
+	println(a)
+	println(Two{})
+	dump(Two{})
+}

